src/Server/Helper.php

```diff
@@ -533,10 +533,10 @@ public function parsePsrRequest(RequestInterface $request)
                     : json_decode((string) $request->getBody(), true);
 
                 if ($bodyParams === null) {
-                    throw new InvariantViolation(
+                    throw new RequestError(
                         $request instanceof ServerRequestInterface
-                         ? 'Expected to receive a parsed body for "application/json" PSR-7 request but got null'
-                         : 'Expected to receive a JSON array in body for "application/json" PSR-7 request'
+                         ? 'Expected to receive a parsed body for "application/json" PSR-7 request, got: null'
```

Comment:
Codecov complains about missing coverage for this line. Now that I think about it more, we can just delete the conditional. Given this is a client facing error, we can simplify to just always say:

```
Expected to receive a JSON body for "application/json" request, got: null
```

More useful for clients, since they do not care about PSR-7 or any other server-internal mechanics.

Reply:
@spawnia IIRC the same payload is represented differently in `ServerRequestInterface`, that's why there's "parsed" body. `ServerRequestInterface` has `getParsedBody()` method.

So in once case it's a json body and in second case it's parsed payload

Reply:
At this point we know the client sent `content-type: application/json`. Thus, I think we should inform them that they should also send a non-null JSON body in their request. The internal method name of the PSR interface seems irrelevant for them?

Reply:
Since we support `ServerRequestInterface` I can't imagine every possible way to construct it so I could tell there was a json at the beginning in all cases. I'd prefer to keep those two branches separated for clarity, though we can definitely reword it, e.g.

```diff
- Expected to receive a parsed body for "application/json" PSR-7 request but got null
+ Expected to receive a body for "application/json" PSR-7 request but the body is missing
```

Reply:
Please read the surrounding code. It must have been JSON, otherwise this execution branch is never entered.